File: backend/routes/company.py
from flask import Blueprint, jsonify, request, session
from models.job import JobPosition
from models.company import Company
from utils.decorators import role_required
from extensions import db
from models.application import Application
from models.student import Student
from extensions import cache
from datetime import datetime
from dateutil import parser   
from models.user import User

company_bp = Blueprint("company", __name__)

# ---------------- Company dashboard --------------------------------
@company_bp.route("/dashboard")
@role_required("company")
def company_dashboard():
    company = Company.query.filter_by(user_id=session["user_id"]).first()

    if not company:
        return jsonify({"message": "Company not found"}), 404

    # total jobs
    total_jobs = JobPosition.query.filter_by(company_id=company.id).count()

    # total active jobs
    active_jobs = JobPosition.query.filter_by(
        company_id=company.id,
        status="Active"
    ).count()

    # total applications (join query)
    total_applications = db.session.query(Application)\
        .join(JobPosition, Application.drive_id == JobPosition.id)\
        .filter(JobPosition.company_id == company.id)\
        .count()

    # Shortlisted count
    shortlisted_count = db.session.query(Application)\
        .join(JobPosition, Application.drive_id == JobPosition.id)\
        .filter(
            JobPosition.company_id == company.id,
            Application.status == "Shortlisted"
        ).count()

    return jsonify({
        "is_approved": company.approved,
        "total_jobs": total_jobs,
        "active_jobs": active_jobs,
        "total_applications": total_applications,
        "shortlisted_count": shortlisted_count
    })




# ---------------------- Create Jobs ---------------------------------
@company_bp.route("/create-job", methods=['POST'])
@role_required("company")
def create_job():
    data = request.json 

    company = Company.query.filter_by(user_id=session["user_id"]).first()
    if not company:
        return jsonify({"message": "Company not found"}), 404
    if not company.approved:
        return jsonify({"message": "Company not approved"}), 403

    job = JobPosition(
        company_id=company.id,
        title = data.get("title"),
        description = data.get("description"),
        salary = data.get("salary"),
        skills_required = data.get("skills_required"),
        eligibility = data.get("eligibility"),
        experience_required = data.get("experience_required"),
        benefits = data.get("benefits"),
        location = data.get("location"),
        status="Pending",
        approved=False
    )
    db.session.add(job)
    db.session.commit()
    cache.delete_memoized(view_company_jobs)

    return jsonify({"message": "Job created successfully, waiting for admin approval"})


# ----------------------- View Own Jobs ------------------------------------------
@company_bp.route("/view-jobs")
@role_required("company")
def view_company_jobs():
    company = Company.query.filter_by(user_id=session["user_id"]).first()
    jobs = JobPosition.query.filter_by(company_id=company.id).all()

    result = []
    for j in jobs:
        
        applicants_count = Application.query.filter_by(drive_id=j.id).count()
        
        result.append({
            "id": j.id,
            "title": j.title,
            "salary": j.salary,
            "location": j.location,
            "status": j.status,
            "approved": j.approved,
            "applicants_count": applicants_count  
        })

    return jsonify({
        "jobs": result,
        "company_approved": company.approved
    })


# ------------------------ Close job ----------------------------------------
@company_bp.route("/close-job/<int:job_id>", methods=["PUT"])
@role_required("company")
def close_job(job_id):

    company = Company.query.filter_by(user_id=session["user_id"]).first()

    job = JobPosition.query.filter_by(id=job_id, company_id=company.id).first()

    if not job:
        return jsonify({"message": "Unauthorized or job not found"}), 403

    job.status = "Closed"
    db.session.commit()

    return jsonify({"message": "Job closed"})


# ------------------------------ Reopen Job -----------------------------------------
@company_bp.route("/reopen-job/<int:job_id>", methods=["PUT"])
@role_required("company")
def reopen_job(job_id):
    company = Company.query.filter_by(user_id=session["user_id"]).first()
    job = JobPosition.query.filter_by(id=job_id, company_id=company.id).first()
    if not job:
        return jsonify({"message": "Unauthorized"}), 403
    job.status = "Active"
    db.session.commit()
    return jsonify({"message": "Job reopened"})

# ------------------------- View Applicants per job --------------------------------------
@company_bp.route("/applicants/<int:job_id>")
@role_required("company")
#@cache.cached(timeout=120, query_string=True)
def view_applicants(job_id):
    company = Company.query.filter_by(user_id=session["user_id"]).first()
    job = JobPosition.query.get(job_id)

    if not job or job.company_id != company.id:
        return jsonify({"message": "Unauthorized access"}), 403

    applications = Application.query.filter_by(drive_id=job_id).all()
    
    # total student count for this job
    total_applicants = len(applications)  # Ya .count()

    result = []
    for app in applications:
        student = Student.query.get(app.student_id)
        
        result.append({
            "application_id": app.id,
            "student_name": student.name,
            "student_id": student.id,
            "branch": student.branch,
            "cgpa": student.cgpa,
            "status": app.status,
            "interview_date": app.interview_date,
        
            # "applicants_count": total_applicants > Yeh har row mein same hoga
        })

    # count ko alag se bheja
    return jsonify({
        "applicants": result,
        "applicants_count": total_applicants
    })
from datetime import datetime
from models.placement import Placement

# ---------------------- Update Application status ----------------------------------
@company_bp.route("/update-application/<int:app_id>", methods=["PUT"])
@role_required("company")
def update_application(app_id):

    company = Company.query.filter_by(user_id=session["user_id"]).first()
    if not company:
        return jsonify({"message": "Company not found"}), 404

    application = Application.query.get(app_id)
    if not application:
        return jsonify({"message": "Application not found"}), 404

    job = JobPosition.query.get(application.drive_id)

    if not job or job.company_id != company.id:
        return jsonify({"message": "Unauthorized"}), 403

    data = request.json
    new_status = data.get("status")

    allowed_status = [
        "Applied",
        "Shortlisted",
        "Interview",
        "Selected",
        "Rejected",
        "Placed"
    ]

    if new_status and new_status not in allowed_status:
        return jsonify({"message": "Invalid status"}), 400

    valid_transitions = {
        "Applied": ["Shortlisted", "Rejected", "Interview"],
        "Shortlisted": ["Interview", "Rejected", "Selected"],
        "Interview": ["Selected", "Rejected", "Interview", "Shortlisted"],
        "Selected": ["Placed", "Rejected"],
        "Rejected": [],
        "Placed": []
    }

    current_status = application.status

    if new_status:
        if new_status not in valid_transitions.get(current_status, []):
            return jsonify({
                "message": f"Invalid transition from {current_status} to {new_status}"
            }), 400

        application.status = new_status

    # Interview Date Update
    if data.get("interview_date"):
        application.interview_date = parser.parse(data.get("interview_date"))
        
    # Feedback
    if data.get("feedback"):
        application.feedback = data.get("feedback")

    # Placement Creation
    if application.status == "Selected":

        existing_placement = Placement.query.filter_by(
            student_id=application.student_id,
            position=job.title,
            company_id=company.id
        ).first()

        if existing_placement:
            if data.get("joining_date"):
                existing_placement.joining_date = parser.parse(data.get("joining_date"))

        else:
            joining_date = None
            if data.get("joining_date"):
                joining_date = parser.parse(data.get("joining_date"))

            placement = Placement(
                student_id=application.student_id,
                company_id=company.id,
                position=job.title,
                salary=job.salary,
                joining_date=joining_date
            )
            db.session.add(placement)
        #Commit
        application.status = "Placed" 
        db.session.commit()

        return jsonify({"message": "Application updated successfully"})
    
    db.session.commit()
    return jsonify({"message": "Application updated successfully"})


# ------------ Company can view student profile which applied for job ----------------------------
@company_bp.route("/student-profile/<int:student_id>/<int:job_id>")
@role_required("company")
def view_student_profile(student_id, job_id):

    company = Company.query.filter_by(user_id=session['user_id']).first()

    # Chekc job belong to thast company
    job = JobPosition.query.get(job_id)
    if not job or job.company_id != company.id:
        return jsonify({"message": "Unauthorized"}), 403
    
    # Check student applied to this job
    application = Application.query.filter_by(
        student_id=student_id,
        drive_id=job_id
    ).first()
    
    if not application:
        return jsonify({"message": "Student did not applied to this job"}), 403
    

    student = Student.query.get(student_id)
    user = User.query.get(student.user_id)

    return jsonify({
        "name": student.name,
        "email": user.email,
        "branch": student.branch,
        "cgpa": student.cgpa,
        "skills": student.skills,
        "experience": student.experience,
        "education": student.education,
        "resume": student.resume
    })


# ---------------- UPDATE JOB ------------------------------------------
@company_bp.route("/update-job/<int:job_id>", methods=["PUT"])
@role_required("company")
def update_job(job_id):
    company = Company.query.filter_by(user_id=session["user_id"]).first()
    job = JobPosition.query.filter_by(id=job_id, company_id=company.id).first()
    if not job:
        return jsonify({"message":"Unauthorized"}),403

    data = request.json
    job.title = data.get("title", job.title)
    job.salary = data.get("salary", job.salary)
    job.skills_required = data.get("skills_required", job.skills_required)
    job.location = data.get("location", job.location)
    job.description = data.get("description", job.description)

    db.session.commit()
    return jsonify({"message":"Job updated"})

File: frontend/static/js/company.js
// ===============================
// COMPANY DASHBOARD - 
// ===============================

const CompanyDashboard = {
    template: `
    <div class="container mt-4">
        <div class="d-flex justify-content-between align-items-center mb-4">
            <h2>Company Dashboard</h2>
            <div>
                <button class="btn btn-outline-primary" :class="{active: tab==='dashboard'}" @click="tab='dashboard'">Dashboard</button>
                <button class="btn btn-outline-primary" :class="{active: tab==='jobs'}" @click="tab='jobs'">My Jobs</button>
                <button class="btn btn-outline-primary" :class="{active: tab==='post'}" @click="tab='post'">Post Job</button>
                <button class="btn btn-outline-primary" :class="{active: tab==='applicants'}" @click="tab='applicants'">Applicants</button>
            </div>
        </div>

        <div v-if="!isApproved" class="alert alert-warning">
            Your company is pending admin approval. Jobs you post will be visible only after approval.
        </div>

        <div v-if="tab==='dashboard'" class="row">
            <div class="col-md-3">
                <div class="card text-center p-3">
                    <h6>Total Jobs</h6>
                    <h3>[[ dashboardStats.total_jobs || 0 ]]</h3>
                </div>
            </div>
            <div class="col-md-3">
                <div class="card text-center p-3">
                    <h6>Active Jobs</h6>
                    <h3>[[ dashboardStats.active_jobs || 0 ]]</h3>
                </div>
            </div>
            <div class="col-md-3">
                <div class="card text-center p-3">
                    <h6>Applications</h6>
                    <h3>[[ dashboardStats.total_applications || 0 ]]</h3>
                </div>
            </div>
            <div class="col-md-3">
                <div class="card text-center p-3">
                    <h6>Shortlisted</h6>
                    <h3>[[ dashboardStats.shortlisted_count || 0 ]]</h3>
                </div>
            </div>
        </div>

        <div v-if="tab==='jobs'">
            <h4>My Job Postings</h4>
            <table class="table table-bordered">
                <thead>
                    <tr>
                        <th>Title</th><th>Salary</th><th>Location</th>
                        <th>Status</th><th>Approval</th><th>Applicants</th><th>Action</th>
                    </tr>
                </thead>
                <tbody>
                    <tr v-for="j in jobs" :key="j.id">
                        <td>[[ j.title ]]</td>
                        <td> [[ j.salary ]] LPA</td>
                        <td>[[ j.location || 'N/A' ]]</td>
                        <td>
                            <span class="badge" :class="{
                                'bg-success': j.status==='Active',
                                'bg-warning': j.status==='Pending',
                                'bg-secondary': j.status==='Closed'
                            }">[[ j.status ]]</span>
                        </td>
                        <td>
                            <span class="badge" :class="j.approved ? 'bg-success' : 'bg-warning'">
                                [[ j.approved ? 'Approved' : 'Pending' ]]
                            </span>
                        </td>
                        <td>
                            <button class="btn btn-sm btn-info" @click="viewApplicants(j.id)">[[ j.applicants_count || 0 ]] View</button>
                        </td>
                        <td>
                            <button v-if="j.status!=='Closed'" class="btn btn-sm btn-warning" @click="closeJob(j.id)">Close</button>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>

        <div v-if="tab==='post'" class="row">
            <div class="col-md-8 mx-auto">
                <div class="card">
                    <div class="card-header">Post New Job / Placement Drive</div>
                    <div class="card-body">
                        <div class="mb-3">
                            <label>Job Title *</label>
                            <input v-model="newJob.title" class="form-control" required>
                        </div>
                        <div class="mb-3">
                            <label>Description *</label>
                            <textarea v-model="newJob.description" class="form-control" rows="3" required></textarea>
                        </div>
                        <div class="row">
                            <div class="col-md-6 mb-3">
                                <label>Salary (LPA) *</label>
                                <input v-model.number="newJob.salary" type="number" step="0.1" class="form-control" required>
                            </div>
                            <div class="col-md-6 mb-3">
                                <label>Location</label>
                                <input v-model="newJob.location" class="form-control">
                            </div>
                        </div>
                        <div class="mb-3">
                            <label>Skills Required *</label>
                            <input v-model="newJob.skills_required" class="form-control" placeholder="Python, SQL, JavaScript" required>
                        </div>
                        <div class="mb-3">
                            <label>Experience Required</label>
                            <input v-model="newJob.experience_required" class="form-control" placeholder="e.g., 0-2 years">
                        </div>
                        <div class="mb-3">
                            <label>Eligibility Criteria</label>
                            <input v-model="newJob.eligibility" class="form-control" placeholder="e.g., CGPA > 7, CSE only">
                        </div>
                        <div class="mb-3">
                            <label>Benefits</label>
                            <textarea v-model="newJob.benefits" class="form-control" rows="2"></textarea>
                        </div>
                        <button class="btn btn-success" @click="submitJob" :disabled="submitting">
                            [[ submitting ? 'Posting...' : 'Post Job' ]]
                        </button>
                    </div>
                </div>
            </div>
        </div>

        <div v-if="tab==='applicants' && !selectedJob" class="text-center text-muted py-5">
            <p>Select a job from the <strong>My Jobs</strong> tab to view its applicants.</p>
            <button class="btn btn-outline-primary btn-sm" @click="tab='jobs'">Go to My Jobs</button>
        </div>

        <div v-if="tab==='applicants' && selectedJob">
            <div class="d-flex justify-content-between mb-3">
                <h4>Applicants for: [[ selectedJob.title ]]</h4>
                <button class="btn btn-secondary btn-sm" @click="tab='jobs'">← Back</button>
            </div>
            
            <table class="table table-bordered">
                <thead>
                    <tr>
                        <th>Student</th><th>Branch</th><th>CGPA</th>
                        <th>Profile</th><th>Status</th><th>Interview Date</th><th>Action</th>
                    </tr>
                </thead>
                <tbody>
                    <tr v-for="a in applicants" :key="a.application_id">
                        <td>[[ a.student_name ]]</td>
                        <td>[[ a.branch ]]</td>
                        <td>[[ a.cgpa ]]</td>
                        <td>
                            <button class="btn btn-sm btn-info" @click="viewStudentProfile(a.student_id)">View</button>
                        </td>
                        <td>
                            <span class="badge" :class="{
                                'bg-primary': a.status==='Applied',
                                'bg-success': a.status==='Shortlisted' || a.status==='Selected' || a.status==='Placed',
                                'bg-info': a.status==='Interview',
                                'bg-danger': a.status==='Rejected'
                            }">[[ a.status ]]</span>
                        </td>
                        <td>
                            <input type="datetime-local" class="form-control form-control-sm" 
                                   v-model="a.interview_date" 
                                   :disabled="a.status==='Placed' || a.status==='Rejected'"
                                   @change="scheduleInterview(a.application_id, a.interview_date)">
                        </td>
                        <td>
                            <select 
                                class="form-select form-select-sm"
                                :disabled="a.status==='Placed'"
                                @change="updateStatus(a, $event.target.value)"
                            >
                                <option value="">Change Status</option>
                                <option value="Shortlisted">Shortlist</option>
                                <option value="Interview">Schedule Interview</option>
                                <option value="Selected">Select & Finalize</option>
                                <option value="Rejected">Reject</option>
                            </select>
                            <textarea v-if="a.status==='Rejected'" class="form-control form-control-sm mt-1" 
                                      placeholder="Feedback" v-model="a.feedback" @blur="saveFeedback(a)"></textarea>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>

        <div v-if="showProfileModal" class="modal fade show" style="display:block; background:rgba(0,0,0,0.5);">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <h5>Student Profile: [[ profileStudent.name ]]</h5>
                        <button type="button" class="btn-close" @click="showProfileModal=false"></button>
                    </div>
                    <div class="modal-body">
                        <p><strong>Branch:</strong> [[ profileStudent.branch ]]</p>
                        <p><strong>CGPA:</strong> [[ profileStudent.cgpa ]]</p>
                        <p><strong>Skills:</strong> [[ profileStudent.skills ]]</p>
                        <p><strong>Education:</strong> [[ profileStudent.education ]]</p>
                        <p><strong>Experience:</strong> [[ profileStudent.experience ]]</p>
                        <p><strong>Email:</strong> [[ profileStudent.email ]]</p>
                        <p v-if="profileStudent.resume">
                            <strong>Resume:</strong>
                            <a :href="profileStudent.resume" target="_blank" class="btn btn-sm btn-outline-primary">
                                Download Resume
                            </a>
                        </p>
                    </div>
                </div>
            </div>
        </div>
    </div>
    `,

    data() {
        return {
            tab: 'dashboard',
            jobs: [],
            applicants: [],
            selectedJob: null,
            dashboardStats: {},
            isApproved: false,
            submitting: false,
            showProfileModal: false,
            profileStudent: {},
            newJob: {
                title: '',
                description: '',
                salary: '',
                location: '',
                skills_required: '',
                experience_required: '',
                eligibility: '',
                benefits: ''
            }
        };
    },

    methods: {
        async loadDashboard() {
            const res = await axios.get('/api/company/dashboard');
            this.dashboardStats = res.data;
            this.isApproved = res.data.is_approved;
        },

        async loadJobs() {
            const res = await axios.get('/api/company/view-jobs');
            this.jobs = res.data.jobs;        
            this.isApproved = res.data.company_approved;
        },

        async submitJob() {
            this.submitting = true;
            try {
                await axios.post('/api/company/create-job', this.newJob);
                alert('Job posted successfully! Waiting for admin approval.');
                this.newJob = { title: '', description: '', salary: '', location: '', skills_required: '', experience_required: '', eligibility: '', benefits: '' };
                this.loadJobs();
                this.tab = 'jobs';
            } catch (err) {
                alert('Failed to post job');
            }
            this.submitting = false;
        },

        async viewApplicants(jobId) {
            this.selectedJob = this.jobs.find(j => j.id === jobId);
            const res = await axios.get(`/api/company/applicants/${jobId}`);
            this.applicants = res.data.applicants;
            this.tab = 'applicants';
        },

        async viewStudentProfile(studentId) {
            const res = await axios.get(`/api/company/student-profile/${studentId}/${this.selectedJob.id}`);
            this.profileStudent = res.data;
            this.showProfileModal = true;
        },

        async updateStatus(applicant, newStatus) {
            if (!newStatus) return;

            try {
                let payload = { status: newStatus };

                if (newStatus === 'Selected') {
                    const date = prompt('Enter joining date (YYYY-MM-DD):');
                    if (!date) {
                        alert('Joining date required');
                        return;
                    }
                    payload.joining_date = date;
                }

                await axios.put(
                    `/api/company/update-application/${applicant.application_id}`,
                    payload
                );

                alert(`Status updated to ${newStatus}`);

                const res = await axios.get(
                    `/api/company/applicants/${this.selectedJob.id}`
                );
                this.applicants = res.data.applicants;

            } catch (err) {
                alert(err.response?.data?.message || 'Failed to update status');
            }
        },
        async scheduleInterview(appId, dateTime) {
            if (!dateTime) return;
            
            try {
                await axios.put(`/api/company/update-application/${appId}`, {
                    interview_date: dateTime,
                    status: 'Interview'
                });
                
                alert('Interview scheduled. Email reminder will be sent to student.');
                
                // Refresh applicants
                const res = await axios.get(`/api/company/applicants/${this.selectedJob.id}`);
                this.applicants = res.data.applicants; 
                
            } catch (err) {
                alert('Failed to schedule interview');
            }
        },

        async saveFeedback(applicant) {
            if (!applicant.feedback) return;
            
            await axios.put(`/api/company/update-application/${applicant.application_id}`, {
                feedback: applicant.feedback
            });
        },

        async closeJob(jobId) {
            if (confirm('Close this job? Students will not be able to apply.')) {
                await axios.put(`/api/company/close-job/${jobId}`);
                this.loadJobs();
            }
        }
    },

    mounted() {
        this.loadDashboard();
        this.loadJobs();
    }
};